src/ripple/nodestore/impl/DatabaseShardImp.h

```diff
@@ -0,0 +1,164 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_DATABASESHARDIMP_H_INCLUDED
+#define RIPPLE_NODESTORE_DATABASESHARDIMP_H_INCLUDED
+
+#include <ripple/nodestore/DatabaseShard.h>
+#include <ripple/nodestore/impl/Shard.h>
+
+namespace ripple {
+namespace NodeStore {
+
+class DatabaseShardImp : public DatabaseShard
+{
+public:
+    DatabaseShardImp() = delete;
+    DatabaseShardImp(DatabaseShardImp const&) = delete;
+    DatabaseShardImp& operator=(DatabaseShardImp const&) = delete;
+
+    DatabaseShardImp(Application& app, std::string const& name,
+        Stoppable& parent, Scheduler& scheduler, int readThreads,
+            Section const& config, beast::Journal j);
+
+    ~DatabaseShardImp() override;
+
+    bool
+    init() override;
+
+    boost::optional<std::uint32_t>
+    prepare(std::uint32_t validLedgerSeq) override;
+
+    std::shared_ptr<Ledger>
+    fetchLedger(uint256 const& hash, std::uint32_t seq) override;
+
+    void
+    setStored(std::shared_ptr<Ledger const> const& ledger) override;
+
+    bool
+    hasLedger(std::uint32_t seq) override;
+
+    std::string
+    getCompleteShards() override;
+
+    void
+    validate() override;
+
+    std::string
+    getName() const override
+    {
+        return "shardstore";
+    }
+
+    void
+    for_each(std::function <void(std::shared_ptr<NodeObject>)> f) override
+    {
+        Throw<std::runtime_error>("Shard store import not supported");
+    }
+
+    void
+    import(Database& source) override
+    {
+        Throw<std::runtime_error>("Shard store import not supported");
+    }
+
+    std::int32_t
+    getWriteLoad() const override;
+
+    void
+    store(NodeObjectType type, Blob&& data,
+        uint256 const& hash, std::uint32_t seq) override;
+
+    std::shared_ptr<NodeObject>
+    fetch(uint256 const& hash, std::uint32_t seq) override;
+
+    bool
+    asyncFetch(uint256 const& hash, std::uint32_t seq,
+        std::shared_ptr<NodeObject>& object) override;
+
+    bool
+    copyLedger(std::shared_ptr<Ledger const> const& ledger) override;
+
+    int
+    getDesiredAsyncReadCount(std::uint32_t seq) override;
+
+    float
+    getCacheHitRate() override;
+
+    void
+    tune(int size, int age) override;
+
+    void
+    sweep() override;
+
+private:
+    Application& app_;
+    mutable std::mutex m_;
```

Comment:
I don't see all member functions taking this lock (`init`, `prepare`). Do only some members need protection?

Reply:
`prepare` takes the lock. There are a couple of private functions that assume the lock is held and that is noted in the comments. `init` doesn't as it should only be called once and before anything else. I will add a throw if erroneously called more than once. Similar thing for `validate`.

Reply:
👍   One option that I learned from @scottschurr  is to have those other private functions take the lock by reference to make it clearer for readers that the lock was already held.  I trust the code more than the comments ;). Not necessary, just for your consideration.

Reply:
Thanks for the suggestion, will add.